ethereum/extethclient/eth_wallet_client.go

```diff
@@ -298,6 +301,30 @@ func (c *ethClient) Raw() *ethclient.Client {
 	return c.ec
 }
 
+func (c *ethClient) CancelTxWithNonce(
+	ctx context.Context,
+	nonce uint64,
+	gasPrice *big.Int,
+) (ethcommon.Hash, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	tx := ethtypes.NewTransaction(nonce, c.ethAddress, big.NewInt(0), 21000, gasPrice, nil)
+
+	signer := ethtypes.LatestSignerForChainID(c.chainID)
+	signedTx, err := ethtypes.SignTx(tx, signer, c.ethPrivKey)
+	if err != nil {
+		return ethcommon.Hash{}, fmt.Errorf("failed to sign tx: %w", err)
+	}
+
+	err = c.ec.SendTransaction(ctx, signedTx)
+	if err != nil {
+		return ethcommon.Hash{}, fmt.Errorf("failed to send tx: %w", err)
+	}
```

Comment:
I see that we are waiting for the receipt in the caller, but maybe we should do it here? (And in the Transfer method as well?) It hasn't really been cancelled until the tx is included in a block. In the monero wallet methods we are waiting for the transactions to be mined. If multiple callers would will all be writing the exact same boilerplate code for the receipt, that would be an argument for doing it here. If you think different callers will want distinct receipt handling, the way it is now would be better.

Reply:
yeah I wrote it as is because the callers might want to handle the receipt checking differently

Reply:
Feel free to disagree, but I think in this case we'd be best off being consistent and waiting for the receipt here. Then, when a secondary use case requiring different behavior becomes apparent, we can do the necessary code refactoring. There is a high probability that the alternate handing would be done completely in this package to promote maximum code reuse. I've spent so many hours of my life tracking down race conditions in the code base because receipts were not being waited on, that I'm in favor of making it hard for anyone accidentally not do it.  :)

Reply:
I'm actually going to argue we need it to be separate here. there may not actually ever be a receipt for the cancellation transaction in the case `newSwap` is executed, as the tx will simply be dropped from the mempool (and same for the newSwap tx in the case the cancellation is included) it would be better here to have a select statement that chooses between which of the two receipts is received, then check the status from there. I don't think the `WaitForReceipt` function should be used here at all, as only one of the two txs will ever be included, and thus one will return `ethereum.NotFound` and then loop inside `WaitForReceipt` for one hour :/

Reply:
I think I need to actually remove usage of `WaitForReceipt` here entirely because of that actually